liblangutil/Scanner.cpp

```diff
@@ -271,12 +273,67 @@ bool Scanner::skipWhitespaceExceptUnicodeLinebreak()
 	return sourcePos() != startPosition;
 }
 
+struct Scanner::UnicodeDirectionOverrideProcessor
+{
+	Scanner& scanner;
+	int rtlOverrideDepth = 0;
+
+	pair<bool, ScannerError> operator()()
+	{
+		// U+202D (LRO - Left-to-Right Override)
+		// U+202E (RLO - Right-to-Left Override)
+		if (
+			scanner.tryScanByteSequence("\xE2\x80\xAD") ||
+			scanner.tryScanByteSequence("\xE2\x80\xAE")
+		)
+		{
+			rtlOverrideDepth++;
+			return pair{true, ScannerError::NoError};
+		}
+		else if (scanner.tryScanByteSequence("\xE2\x80\xAC")) // U+202C (PDF - Pop Directional Formatting)
+		{
+			rtlOverrideDepth--;
+			if (rtlOverrideDepth < 0)
+				return pair{true, ScannerError::DirectionalOverrideUnderflowInComment};
```

Comment:
The [spec](https://unicode.org/reports/tr9/#Terminating_Explicit_Directional_Embeddings_and_Overrides) indicates that PDF can follow also LRE and RLE in valid text. This code would return an error in such cases.

Reply:
Also "Terminating Isolates" in [section 3.3.2 Explicit Levels and Directions](https://unicode.org/reports/tr9/#Explicit_Levels_and_Directions) says:
> A PDI terminates the scope of the isolate initiator it matches. It also terminates the scopes of all embedding initiators within the scope of the matched isolate initiator for which a matching PDF has not been encountered. If it does not match any isolate initiator, it is ignored.

So I guess, we should count a PDI as a terminator too (as long as there's an unterminated RLE/LRE) even if we do not support LRI/RLI.

Reply:
And from "End of Paragraph" in that section:
> X8. All explicit directional embeddings, overrides and isolates are completely terminated at the end of each paragraph.

Which, I think means that we should account for any [LS](https://www.fileformat.info/info/unicode/char/2028/) and [PS](https://www.fileformat.info/info/unicode/char/2029/) characters too?

Reply:
Can LRO/RLO follow eachother without a PDF? e.g. is <LRO>...<RLO>...<PDF> valid?

Reply:
@axic `<LRO> ... <RLO> ... <PDF> <PDF>` this works, yes. you could treat an `LRO` and `RLO` as push, and both therefore have to be popped.

@cameel I will be adding isolation marker counting, too. (tomorrow).

But with regards to "end of paragraph", I was reading that too, and I think our end-of-paragraph can simply be the end of the comment, no need to be "fully" validating unicode spec here unless it helps improving security. :)

Reply:
> no need to be "fully" validating unicode spec here unless it helps improving security. :)

Agreed. But we also need to be careful to avoid rejecting stuff that's just normal in bidi languages. Encountering a paragraph separator in the middle of a comment line probably is not but I don't write in any such language so I'm just being cautious :)

Reply:
@cameel I may be knowing one, I hope i can catch him next week. but I think there's no harm in being not too restrictive (also, we are not an editor / text render :), but only want to guard against certain attacks in source code)